Constraint on multi-column property causes hibernate mapping failure

Upgrading from Grails 2.5.5 to 3.3.2 and domains where we had a constraint on a multi-column property are causing an error on startup.

Hibernate 5.1.9
Gorm 6.1.8
Grails 3.3.2
Java 8

This used to work properly. The nullable constraint would be applied to each of the underlying columns.

For example:
```
import org.jadira.usertype.dateandtime.joda.PersistentDateTimeAndZoneWithOffset
import org.joda.time.DateTime
import org.joda.time.LocalDate

class Author {
	String name
	DateTime testDate

	static constraints = {
		testDate(nullable: true)
	}

	static mapping = {
	    testDate(type: PersistentDateTimeAndZoneWithOffset) {
	        column name: "testDate_datetime"
	        column name: "testDate_offset"
	    }
	}
}
```

Results in:
```
Caused by: java.lang.RuntimeException: Cannot treat multi-column property as a single-column property
        at org.grails.orm.hibernate.cfg.PropertyConfig.checkHasSingleColumn(PropertyConfig.groovy:449)
        at org.grails.orm.hibernate.cfg.PropertyConfig.getScale(PropertyConfig.groovy:425)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.MethodMetaProperty$GetBeanMethodMetaProperty.getProperty(MethodMetaProperty.java:76)
        at org.codehaus.groovy.runtime.callsite.GetEffectivePogoPropertySite.callGroovyObjectGetProperty(GetEffectivePogoPropertySite.java:68)
        at org.grails.orm.hibernate.cfg.HibernateMappingBuilder$_closure1.doCall(HibernateMappingBuilder.groovy:478)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
        at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:925)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:908)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodN(ScriptBytecodeAdapter.java:168)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeClosure(ScriptBytecodeAdapter.java:574)
        at org.grails.orm.hibernate.cfg.HibernateMappingBuilder.methodMissing(HibernateMappingBuilder.groovy:691)
```
Removing the constraints reference fixes the error, but then both of the properties on the domain object are required and non-nullable.

It's possible this is a by design change, but I can't really find any reference to how the two interact.

--- a/grails-datastore-gorm-hibernate-core/src/main/groovy/org/grails/orm/hibernate/cfg/HibernateMappingBuilder.groovy
+++ b/grails-datastore-gorm-hibernate-core/src/main/groovy/org/grails/orm/hibernate/cfg/HibernateMappingBuilder.groovy
@@ -475,8 +475,12 @@ class HibernateMappingBuilder implements MappingConfigurationBuilder<Mapping, Pr
             property.max = namedArgs.max instanceof Comparable ? namedArgs.max : property.max
             property.min = namedArgs.min instanceof Comparable ? namedArgs.min : property.min
             property.range = namedArgs.range instanceof ObjectRange ? namedArgs.range : null
-            property.scale = namedArgs.scale instanceof Integer ? namedArgs.scale : property.scale
             property.inList = namedArgs.inList instanceof List ? namedArgs.inList : property.inList
+
+            // Need to guard around calling getScale() for multi-column properties (issue #1048)
+            if (namedArgs.scale instanceof Integer) {
+                property.scale = (Integer)namedArgs.scale
+            }
 
             if (namedArgs.fetch) {
                 switch(namedArgs.fetch) {

--- a/grails-datastore-gorm-hibernate-core/src/test/groovy/grails/gorm/hibernate/mapping/HibernateMappingBuilderTests.groovy
+++ b/grails-datastore-gorm-hibernate-core/src/test/groovy/grails/gorm/hibernate/mapping/HibernateMappingBuilderTests.groovy
@@ -5,6 +5,7 @@ package grails.gorm.hibernate.mapping
  */
 import org.grails.orm.hibernate.cfg.CompositeIdentity
 import org.grails.orm.hibernate.cfg.HibernateMappingBuilder
+import org.grails.orm.hibernate.cfg.Mapping
 import org.grails.orm.hibernate.cfg.PropertyConfig
 import org.hibernate.FetchMode
 
@@ -692,6 +693,43 @@ class HibernateMappingBuilderTests extends GroovyTestCase {
 
         shouldFail { mapping.columns.amount.column }
         shouldFail { mapping.columns.amount.sqlType }
+    }
+
+    void testConstrainedPropertyWithMultipleColumns() {
+        def builder = new HibernateMappingBuilder("Foo")
+        builder.evaluate {
+            amount type: MyUserType, {
+                column name: "value"
+                column name: "currency", sqlType: "char", length: 3
+            }
+        }
+        def mapping = builder.evaluate {
+            amount nullable: true
+        }
+
+        assertEquals 2, mapping.columns.amount.columns.size()
+        assertEquals "value", mapping.columns.amount.columns[0].name
+        assertEquals "currency", mapping.columns.amount.columns[1].name
+        assertEquals "char", mapping.columns.amount.columns[1].sqlType
+        assertEquals 3, mapping.columns.amount.columns[1].length
+
+        shouldFail { mapping.columns.amount.column }
+        shouldFail { mapping.columns.amount.sqlType }
+    }
+
+    void testDisallowedConstrainedPropertyWithMultipleColumns() {
+        def builder = new HibernateMappingBuilder("Foo")
+        builder.evaluate {
+            amount type: MyUserType, {
+                column name: "value"
+                column name: "currency", sqlType: "char", length: 3
+            }
+        }
+        shouldFail {
+            builder.evaluate {
+                amount scale: 2
+            }
+        } == "Cannot treat multi-column property as a single-column property"
     }
 
     void testPropertyWithUserTypeAndNoParams() {

--- a/grails-datastore-gorm-hibernate-core/src/test/groovy/grails/gorm/hibernate/mapping/HibernateMappingBuilderTests.groovy
+++ b/grails-datastore-gorm-hibernate-core/src/test/groovy/grails/gorm/hibernate/mapping/HibernateMappingBuilderTests.groovy
@@ -5,7 +5,6 @@ package grails.gorm.hibernate.mapping
  */
 import org.grails.orm.hibernate.cfg.CompositeIdentity
 import org.grails.orm.hibernate.cfg.HibernateMappingBuilder
-import org.grails.orm.hibernate.cfg.Mapping
 import org.grails.orm.hibernate.cfg.PropertyConfig
 import org.hibernate.FetchMode
 
@@ -725,7 +724,7 @@ class HibernateMappingBuilderTests extends GroovyTestCase {
                 column name: "currency", sqlType: "char", length: 3
             }
         }
-        shouldFail {
+        assert shouldFail {
             builder.evaluate {
                 amount scale: 2
             }

